module/zcommon/zfs_namecheck.c

```diff
@@ -54,15 +51,89 @@
  */
 int zfs_max_dataset_nesting = 50;
 
-static int
-valid_char(char c)
+/*
+ * Define our own character classes to avoid mistakes with locale.
+ */
+static inline int
+valid_lower(const char c)
+{
+	return (c >= 'a' && c <= 'z');
+}
+
+static inline int
+valid_upper(const char c)
 {
-	return ((c >= 'a' && c <= 'z') ||
-	    (c >= 'A' && c <= 'Z') ||
-	    (c >= '0' && c <= '9') ||
+	return (c >= 'A' && c <= 'Z');
+}
+
+static inline int
+valid_alpha(const char c)
+{
+	return (valid_lower(c) || valid_upper(c));
+}
+
+static inline int
+valid_digit(const char c)
+{
+	return (c >= '0' && c <= '9');
+}
+
+static inline int
+valid_char(const char c)
+{
+	return (valid_alpha(c) || valid_digit(c) ||
 	    c == '-' || c == '_' || c == '.' || c == ':' || c == ' ');
 }
 
+static int
+disklike(const char *s)
+{
+	/*
+	 * Non-exhaustive disk name patterns for various platforms with ZFS
+	 *
+	 * FreeBSD: da# md# ada# nda# nvd# mfid# vtbd#
+	 * Linux: sdA#
+	 * NetBSD: dk# sd# wd#
+	 * Solaris: c#t#d#
```

Comment:
i think we can replace Solaris keyword by 'illumos and dilos' - they are both known and Open Sources